RestEasy Reactive Multipart response fails with files > 2GB

### Describe the bug

large response with file > 2GB fails

### Expected behavior

_No response_

### Actual behavior

An exception is thrown:
```
10:11:13,961 INFO  [app] 10:11:13,532 Request failed: java.lang.RuntimeException: java.lang.OutOfMemoryError: Required array length 2147483639 + 9 is too large
10:11:13,962 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:242)
10:11:13,962 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:178)
10:11:13,962 INFO  [app]        at org.jboss.resteasy.reactive.server.core.serialization.FixedEntityWriter.write(FixedEntityWriter.java:26)
10:11:13,962 INFO  [app]        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:33)
10:11:13,962 INFO  [app]        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:15)
10:11:13,962 INFO  [app]        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
10:11:13,962 INFO  [app]        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:545)
10:11:13,962 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
10:11:13,962 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
10:11:13,962 INFO  [app]        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
10:11:13,962 INFO  [app]        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
10:11:13,962 INFO  [app]        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
10:11:13,962 INFO  [app]        at java.base/java.lang.Thread.run(Thread.java:833)
10:11:13,962 INFO  [app] Caused by: java.lang.OutOfMemoryError: Required array length 2147483639 + 9 is too large
10:11:13,962 INFO  [app]        at java.base/jdk.internal.util.ArraysSupport.hugeLength(ArraysSupport.java:649)
10:11:13,963 INFO  [app]        at java.base/jdk.internal.util.ArraysSupport.newLength(ArraysSupport.java:642)
10:11:13,963 INFO  [app]        at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:100)
10:11:13,963 INFO  [app]        at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:130)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.doWrite(FileBodyHandler.java:66)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.writeTo(FileBodyHandler.java:58)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.writeTo(FileBodyHandler.java:20)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.serialiseEntity(MultipartMessageBodyWriter.java:133)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.write(MultipartMessageBodyWriter.java:84)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeMultiformData(MultipartMessageBodyWriter.java:57)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeResponse(MultipartMessageBodyWriter.java:44)
10:11:13,963 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:207)
10:11:13,963 INFO  [app]        ... 12 more
```

### How to Reproduce?

```
git clone https://github.com/michalszynkiewicz/resteasy-reactive-large-multipart-failure
cd resteasy-reactive-large-multipart-failure
mvn clean install -Dtest=FileTest#downloadMultipart
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

main

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/.github/workflows/jdk-early-access-build.yml
+++ b/.github/workflows/jdk-early-access-build.yml
@@ -24,7 +24,7 @@ env:
   # Workaround testsuite locale issue
   LANG: en_US.UTF-8
   MAVEN_OPTS: -Xmx2g -XX:MaxMetaspaceSize=1g
-  JVM_TEST_MAVEN_OPTS: "-e -B --settings .github/mvn-settings.xml -Dtest-containers -Dstart-containers -Dformat.skip"
+  JVM_TEST_MAVEN_OPTS: "-e -B --settings .github/mvn-settings.xml -Dtest-containers -Dstart-containers -Dtest-resteasy-reactive-large-files -Dformat.skip"
   DB_USER: hibernate_orm_test
   DB_PASSWORD: hibernate_orm_test
   DB_NAME: hibernate_orm_test

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartOutputResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartOutputResource.java
@@ -1,6 +1,8 @@
 package io.quarkus.resteasy.reactive.server.test.multipart;
 
 import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
 import java.util.List;
 
 import javax.ws.rs.GET;
@@ -19,6 +21,7 @@ public class MultipartOutputResource {
     public static final List<String> RESPONSE_VALUES = List.of("one", "two");
     public static final boolean RESPONSE_ACTIVE = true;
 
+    private static final long ONE_GIGA = 1024l * 1024l * 1024l * 1l;
     private final File TXT_FILE = new File("./src/test/resources/lorem.txt");
 
     @GET
@@ -57,10 +60,25 @@ public class MultipartOutputResource {
     @GET
     @Path("/with-file")
     @Produces(MediaType.MULTIPART_FORM_DATA)
-    public MultipartOutputFileResponse complex() {
+    public MultipartOutputFileResponse file() {
         MultipartOutputFileResponse response = new MultipartOutputFileResponse();
         response.name = RESPONSE_NAME;
         response.file = TXT_FILE;
+        return response;
+    }
+
+    @GET
+    @Path("/with-large-file")
+    @Produces(MediaType.MULTIPART_FORM_DATA)
+    public MultipartOutputFileResponse largeFile() throws IOException {
+        File largeFile = File.createTempFile("rr-large-file", ".tmp");
+        largeFile.deleteOnExit();
+        RandomAccessFile f = new RandomAccessFile(largeFile, "rw");
+        f.setLength(ONE_GIGA);
+
+        MultipartOutputFileResponse response = new MultipartOutputFileResponse();
+        response.name = RESPONSE_NAME;
+        response.file = largeFile;
         return response;
     }
 

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartOutputUsingBlockingEndpointsTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartOutputUsingBlockingEndpointsTest.java
@@ -8,6 +8,7 @@ import javax.ws.rs.core.MediaType;
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledIfSystemProperty;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 import io.quarkus.resteasy.reactive.server.test.multipart.other.OtherPackageFormDataBase;
@@ -82,6 +83,16 @@ public class MultipartOutputUsingBlockingEndpointsTest extends AbstractMultipart
 
         assertContainsValue(response, "name", MediaType.TEXT_PLAIN, MultipartOutputResource.RESPONSE_NAME);
         assertContainsFile(response, "file", MediaType.APPLICATION_OCTET_STREAM, "lorem.txt");
+    }
+
+    @EnabledIfSystemProperty(named = "test-resteasy-reactive-large-files", matches = "true")
+    @Test
+    public void testWithLargeFiles() {
+        RestAssured.given()
+                .get("/multipart/output/with-large-file")
+                .then()
+                .contentType(ContentType.MULTIPART)
+                .statusCode(200);
     }
 
     @Test

--- a/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/multipart/MultipartResponseTest.java
+++ b/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/multipart/MultipartResponseTest.java
@@ -8,6 +8,7 @@ import java.io.File;
 import java.io.FileOutputStream;
 import java.io.FileReader;
 import java.io.IOException;
+import java.io.RandomAccessFile;
 import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.time.Duration;
@@ -29,6 +30,7 @@ import org.jboss.resteasy.reactive.MultipartForm;
 import org.jboss.resteasy.reactive.PartType;
 import org.jboss.resteasy.reactive.RestForm;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledIfSystemProperty;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 import io.quarkus.test.QuarkusUnitTest;
@@ -38,6 +40,8 @@ import io.smallrye.mutiny.Uni;
 public class MultipartResponseTest {
 
     public static final String WOO_HOO_WOO_HOO_HOO = "Woo hoo, woo hoo hoo";
+    private static final long ONE_GIGA = 1024l * 1024l * 1024l * 1l;
+
     @TestHTTPResource
     URI baseUri;
 
@@ -96,6 +100,15 @@ public class MultipartResponseTest {
         assertThat(data.panda).isNull();
         assertThat(data.number).isEqualTo(1984);
         assertThat(data.numberz).isNull();
+    }
+
+    @EnabledIfSystemProperty(named = "test-resteasy-reactive-large-files", matches = "true")
+    @Test
+    void shouldParseMultipartResponseWithLargeFile() {
+        Client client = RestClientBuilder.newBuilder().baseUri(baseUri).build(Client.class);
+        MultipartData data = client.getLargeFile();
+        assertThat(data.file).exists();
+        assertThat(data.file.length()).isEqualTo(ONE_GIGA);
     }
 
     @Test
@@ -166,6 +179,11 @@ public class MultipartResponseTest {
 
         @GET
         @Produces(MediaType.MULTIPART_FORM_DATA)
+        @Path("/large")
+        MultipartData getLargeFile();
+
+        @GET
+        @Produces(MediaType.MULTIPART_FORM_DATA)
         @Path("/empty")
         MultipartData getFileEmpty();
 
@@ -193,8 +211,7 @@ public class MultipartResponseTest {
         @GET
         @Produces(MediaType.MULTIPART_FORM_DATA)
         public MultipartData getFile() throws IOException {
-            File file = File.createTempFile("toDownload", ".txt");
-            file.deleteOnExit();
+            File file = createTempFileToDownload();
             // let's write Woo hoo, woo hoo hoo 10k times
             try (FileOutputStream out = new FileOutputStream(file)) {
                 for (int i = 0; i < 10000; i++) {
@@ -209,12 +226,21 @@ public class MultipartResponseTest {
         @Produces(MediaType.MULTIPART_FORM_DATA)
         @Path("/small")
         public MultipartData getSmallFile() throws IOException {
-            File file = File.createTempFile("toDownload", ".txt");
-            file.deleteOnExit();
+            File file = createTempFileToDownload();
             // let's write Woo hoo, woo hoo hoo 1 time
             try (FileOutputStream out = new FileOutputStream(file)) {
                 out.write(WOO_HOO_WOO_HOO_HOO.getBytes(StandardCharsets.UTF_8));
             }
+            return new MultipartData("foo", file, null, 1984, null);
+        }
+
+        @GET
+        @Produces(MediaType.MULTIPART_FORM_DATA)
+        @Path("/large")
+        public MultipartData getLargeFile() throws IOException {
+            File file = createTempFileToDownload();
+            RandomAccessFile f = new RandomAccessFile(file, "rw");
+            f.setLength(ONE_GIGA);
             return new MultipartData("foo", file, null, 1984, null);
         }
 
@@ -230,6 +256,12 @@ public class MultipartResponseTest {
         @Path("/error")
         public MultipartData throwError() {
             throw new RuntimeException("forced error");
+        }
+
+        private static File createTempFileToDownload() throws IOException {
+            File file = File.createTempFile("toDownload", ".txt");
+            file.deleteOnExit();
+            return file;
         }
     }
 

--- a/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientSerialisers.java
+++ b/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientSerialisers.java
@@ -1,7 +1,6 @@
 package org.jboss.resteasy.reactive.client.impl;
 
 import io.vertx.core.buffer.Buffer;
-import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
@@ -104,10 +103,10 @@ public class ClientSerialisers extends Serialisers {
 
         if (writer.isWriteable(entityClass, entityType, entity.getAnnotations(), entity.getMediaType())) {
             if (writerInterceptors == null) {
-                ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                VertxBufferOutputStream out = new VertxBufferOutputStream();
                 writer.writeTo(entityObject, entityClass, entityType, entity.getAnnotations(),
-                        entity.getMediaType(), headerMap, baos);
-                return Buffer.buffer(baos.toByteArray());
+                        entity.getMediaType(), headerMap, out);
+                return out.getBuffer();
             } else {
                 return runClientWriterInterceptors(entityObject, entityClass, entityType, entity.getAnnotations(),
                         entity.getMediaType(), headerMap, writer, writerInterceptors, properties, serialisers,

--- a/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/VertxBufferOutputStream.java
+++ b/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/VertxBufferOutputStream.java
@@ -1,0 +1,32 @@
+package org.jboss.resteasy.reactive.client.impl;
+
+import io.vertx.core.buffer.Buffer;
+import java.io.IOException;
+import java.io.OutputStream;
+
+public class VertxBufferOutputStream extends OutputStream {
+    private Buffer buffer;
+
+    public VertxBufferOutputStream() {
+        this.buffer = Buffer.buffer();
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        buffer.appendByte((byte) (b & 0xFF));
+    }
+
+    @Override
+    public void write(byte[] b) throws IOException {
+        buffer.appendBytes(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        buffer.appendBytes(b, off, len);
+    }
+
+    public Buffer getBuffer() {
+        return this.buffer.copy();
+    }
+}

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultipartMessageBodyWriter.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultipartMessageBodyWriter.java
@@ -1,6 +1,5 @@
 package org.jboss.resteasy.reactive.server.core.multipart;
 
-import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.IOException;
 import java.io.OutputStream;
@@ -81,7 +80,7 @@ public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWrite
                 writeLine(outputStream, charset);
 
                 // write content
-                write(outputStream, serialiseEntity(part.getValue(), part.getType(), requestContext));
+                writeEntity(outputStream, part.getValue(), part.getType(), requestContext);
                 // extra line
                 writeLine(outputStream, charset);
             }
@@ -116,7 +115,7 @@ public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWrite
         os.write(LINE_SEPARATOR.getBytes(defaultCharset));
     }
 
-    private byte[] serialiseEntity(Object entity, MediaType mediaType, ResteasyReactiveRequestContext context)
+    private void writeEntity(OutputStream os, Object entity, MediaType mediaType, ResteasyReactiveRequestContext context)
             throws IOException {
         ServerSerialisers serializers = context.getDeployment().getSerialisers();
         Class<?> entityClass = entity.getClass();
@@ -125,13 +124,12 @@ public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWrite
         MessageBodyWriter<Object>[] writers = (MessageBodyWriter<Object>[]) serializers
                 .findWriters(null, entityClass, mediaType, RuntimeType.SERVER)
                 .toArray(ServerSerialisers.NO_WRITER);
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
         boolean wrote = false;
         for (MessageBodyWriter<Object> writer : writers) {
             if (writer.isWriteable(entityClass, entityType, Serialisers.NO_ANNOTATION, mediaType)) {
                 // FIXME: spec doesn't really say what headers we should use here
                 writer.writeTo(entity, entityClass, entityType, Serialisers.NO_ANNOTATION, mediaType,
-                        Serialisers.EMPTY_MULTI_MAP, baos);
+                        Serialisers.EMPTY_MULTI_MAP, os);
                 wrote = true;
                 break;
             }
@@ -140,7 +138,6 @@ public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWrite
         if (!wrote) {
             throw new IllegalStateException("Could not find MessageBodyWriter for " + entityClass + " as " + mediaType);
         }
-        return baos.toByteArray();
     }
 
     private String generateBoundary() {

